File: fixtures/scale/zig/large/src/core/rng.zig
// xorshift32 generator used where a family needs extra entropy
const std = @import("std");

pub const CoreRng = struct {
    state: u32,
};

pub fn coreRngInit(seed: u32) CoreRng {
    return CoreRng{ .state = if (seed == 0) 0xdeadbeef else seed };
}

pub fn coreRngNext(r: *CoreRng) u32 {
    var x = r.state;
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    r.state = x;
    return x;
}

File: fixtures/scale/rust/large/src/core/rng.rs
/// Small deterministic generator so runs are reproducible without
/// pulling in an external crate.
pub struct XorShift32 {
    state: u32,
}

impl XorShift32 {
    pub fn new(seed: u32) -> Self {
        XorShift32 { state: seed.max(1) }
    }

    pub fn next_u32(&mut self) -> u32 {
        let mut x = self.state;
        x ^= x << 13;
        x ^= x >> 17;
        x ^= x << 5;
        self.state = x;
        x
    }

    pub fn fill(&mut self, len: usize) -> Vec<u8> {
        (0..len).map(|_| (self.next_u32() >> 16) as u8).collect()
    }
}
